### ISSUE TITLE
Using load_threaded_request with sub_threads raises errors when requesting multiple scenes with scripts that access the same autoloaded node

### ISSUE THREAD

**Issue by ericdsw** [godotengine/godot#98865]
### Tested versions

Godot 4.3.stable

### System information

Mac OS Sequoia (15.1)

### Issue description

I was trying to implement a node that loads several scenes in the ResourceLoader's cache before the game starts, and I managed to get a crude implementation of what I wanted by following the documentation for ResourceLoader's `load_threaded_request` and `load_threaded_get_status` method.

However, when I tried to set`load_thread_request`'s third parameter (`use_sub_threads`) to true, several errors popped up on the console complaining that I was trying to call a function from a thread.

Trying to narrow down what was causing the issue, I noticed that it only happened under these circumstances:

- Registered some node as an autoload in the project's settings
- Started the threaded loading logic for more than 1 resource
- Said loaded resources tried to call a method from the global inside their scripts
- The errors pop up when the resources are loading, regardless of whether I instantiate them or not.

<img width="1087" alt="Screenshot 2024-11-05 at 4 23 32 PM" src="https://github.com/user-attachments/assets/5dc80fad-8d0a-4282-8cef-7d22fc1778c6">

While the game doesn't crash, I'm not sure if those errors would cause further problems down the road, or if a compiled version of the game would indeed start crashing. 

Another issue I noticed is that only the first call of `load_threaded_get` returns the actual resource, any subsequent call returns `null`. Not sure if this is truly a bug or if there are some additional steps I need to take to query that resource again.

### Steps to reproduce

Just run the attached project, I've hooked the loading logic to the main scene's `_ready` call, you should see the errors pop up on the debugger, and the text `"finished loading"` should be printed on the console.

For the second issue: after that text appears, pressing the button on the screen prints the loaded resource for one specific scene, but any subsequent button press will instead print null.

### Minimal reproduction project (MRP)

[TestResourceLoader.zip](https://github.com/user-attachments/files/17638784/TestResourceLoader.zip)

This project tires to load all the scenes located inside the `Scenes` subfolder, and both `ChildScene.tscn` and `SecondChildScene.tscn` contain a script that tries to call a method from an autoloaded node called `RandomGlobal.tscn`

**Comment by coffeebeats** [comment]
Hey all - I'm encountering this bug as well, but I think the trigger conditions are a little more specific. I've outlined them below, but I'd suggest updating the issue's title to **`Resource.load_threaded_request` with sub threads raises error if scene has script that references an auto-loaded scene`**.

#### Reproduction

Givens:
- An auto-loaded scene, `AutoLoad`
- A script `caller.gd` which _references_ `AutoLoad`
- A scene with a node that attaches `caller.gd`, `scene.tscn`

When calling `ResourceLoader.load_threaded_request(<path/to/scene.tscn>,"",true)` (`use_sub_threads` must be set to `true` to trigger), then this bug will _often_ (but not always) occur.

I've created a GitHub repository (https://github.com/coffeebeats/example-godot-threaded-load-autoload-bug) which demonstrates this exact scenario in action.

#### Observations

- This error does not seem to prevent the loaded scene from running correctly.
- This only occurs if the auto-load is a scene (not a script).
- This error doesn't occur all of the time, so it's likely a race condition.
  - It occurs _nearly all of the time_ if the load request is invoked from a caller's `_ready` callback. Deferring the call to load the scene does not fix the problem or decrease the rate of occurrence under these circumstances.
  - It occurs _rarely_ if invoked as a result of a user input (e.g. a button press) after the loader scene has been ready for a while.

**Comment by hpvb** [comment]
I can not reproduce this issue after #119757 can you please re-check as well?

**Comment by AThousandShips** [comment]
Assuming fixed as per above, if this can still be replicated in 4.7 please comment and this can be reopened

**Comment by slaff23** [comment]
I am still encountering this bug in v4.7.1.stable.official [a13da4feb] and v4.8.dev2 [7220d456d] (Windows 10, Vulkan Forward+).
Bug still occurs when using `use_sub_threads = true` in `load_threaded_request(path, "", true)`.
When `use_sub_threads` is off, no errors at all.

Error appears for multiple scene autoload scripts (error line is from v4.7.1.stable):
> ERROR: /root/*AutoloadName*: The caller thread can't call the function get_script() on this node. Use call_deferred() or call_deferred_thread_group() instead.
> at: get_script (scene/main/node.cpp:4159)

It seems error only occurs for scene autoloads and not for script autoloads as the engine is probably reaching for the code via `get_script`. Is it intended to never use scene autoloads with `sub_threads`?
The load in error occurs only once per process (the first load that compiles those scripts - loading the same scene again in the same run is clean). The error is not every run as I assume it is a schedule dependent if it lands on a sub-worker-thread. But it can be reliably reproduced with multiple runs.

I could not reproduce it in a minimal project so far. It only shows in my full project.
My knowledge of the engine source is too limited to pinpoint the breaking point, but with some prints in a 4.7.1-stable build I can see that a minimal project reaches the same autoload branch and passes. In my project some nested off-main load tasks run with the flag `is_current_thread_safe_for_nodes()` already cleared and since the marking in `core/io/resource_loader.cpp` is gated on `load_nesting == 0` it never gets restored. Maybe something in `core/object/worker_thread_pool.cpp` clears it on a thread that is already inside a load?

Let me know if you have any good idea how I can better support or eventually reproduce the behavior in a minimal project...
